### PR TITLE
Fido.fetch should accept pathlib.Path objects

### LINKED ISSUE #2400: Fido.fetch should accept pathlib.Path objects.
i.e.:

```python
path = Path('/home/stuart/sunpy/data/jsocflare/')
results = Fido.search(...)
files = Fido.fetch(results, path=path)
```

### PR DESCRIPTION
**Pull request by jnanjekye**
This PR fixes #2400.

### CODE REVIEW

**Comment by jnanjekye** [comment]
@Cadair @nabobalis .. I am trying to run tests to validate this PR on my fork but also on master. I am running python 3.6. Could it be a python version thing:
```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/_pytest/config.py", line 379, in _importconftest
    mod = conftestpath.pyimport()
  File "/usr/local/lib/python2.7/dist-packages/py/_path/local.py", line 668, in pyimport
    __import__(modname)
  File "/usr/local/lib/python2.7/dist-packages/_pytest/assertion/rewrite.py", line 212, in load_module
    py.builtin.exec_(co, mod.__dict__)
  File "/usr/local/lib/python2.7/dist-packages/py/_builtin.py", line 221, in exec_
    exec2(obj, globals, locals)
  File "<string>", line 7, in exec2
  File "/home/programmer/sunpy-git/sunpy/conftest.py", line 16, in <module>
    from sunpy.tests.hash import HASH_LIBRARY_NAME
  File "/home/programmer/sunpy-git/sunpy/tests/hash.py", line 8, in <module>
    import matplotlib.pyplot as plt
ImportError: No module named matplotlib.pyplot
ERROR: could not load /home/programmer/sunpy-git/sunpy/conftest.py
```

**Comment by nabobalis** [comment]
@jnanjekye You aren't running Py3.6 if you get that error. 

```/usr/local/lib/python2.7/dist-packages/_pytest/assertion/rewrite.py```

I suspect you are using system python and not the conda version.

**Comment by jnanjekye** [comment]
hmm okay, let me investigate.

**Comment by nabobalis** [comment]
The PEP8 changes to the long times I strongly disapprove of. They should be changed back.

File: sunpy/database/database.py [hunk 1]
```diff
@@ -384,10 +386,14 @@ def commit(self):
         """
         self.session.commit()
 
-    def _download_and_collect_entries(self, query_result, **kwargs):
+    def _download_and_collect_entries(self, query_result, **kwargs, path=''):
```
Comment (nabobalis) [inline sunpy/database/database.py]: keyword arguments should be before any `**kwargs`

File: sunpy/database/database.py [hunk 2]
```diff
@@ -384,10 +386,14 @@ def commit(self):
...
 
         client = kwargs.pop('client', None)
-        path = kwargs.pop('path', None)
+        if path is None:
+            path = kwargs.pop('path', None)
+        else:
+            path = path
```
Comment (nabobalis) [inline sunpy/database/database.py]: Is this `if,else` needed since path is now a string which is not `None` by default and a argument to the function instead of a kwarg

File: sunpy/database/database.py [hunk 3]
```diff
@@ -454,7 +460,7 @@ def download(self, *query, **kwargs):
 
         return self.fetch(*query, **kwargs)
 
-    def fetch(self, *query, **kwargs):
+    def fetch(self, *query, **kwargs, path=''):
```
Comment (nabobalis) [inline sunpy/database/database.py]: Order of inputs is off here again.

File: sunpy/net/dataretriever/tests/test_eve.py [hunk 4]
```diff
@@ -17,19 +17,26 @@
...
+     SpWx/2012/20120506_EVE_L0CS_DIODES_1m.txt'
      ),
     (TimeRange('2012/7/7', '2012/7/14'),
-     'http://lasp.colorado.edu/eve/data_access/evewebdata/quicklook/L0CS/SpWx/2012/20120707_EVE_L0CS_DIODES_1m.txt',
-     'http://lasp.colorado.edu/eve/data_access/evewebdata/quicklook/L0CS/SpWx/2012/20120714_EVE_L0CS_DIODES_1m.txt'
+     'http://lasp.colorado.edu/eve/data_access/evewebdata/quicklook/L0CS/\n\
+     SpWx/2012/20120707_EVE_L0CS_DIODES_1m.txt',
+     'http://lasp.colorado.edu/eve/data_access/evewebdata/quicklook/L0CS/\n\
+     SpWx/2012/20120714_EVE_L0CS_DIODES_1m.txt'
      )
```
Comment (nabobalis) [inline sunpy/net/dataretriever/tests/test_eve.py]: These all should be reverted back.

File: sunpy/net/dataretriever/tests/test_eve.py [hunk 5]
```diff
@@ -82,9 +89,14 @@ def test_fido(query):
...
 
+    path = Path('../database/tests/')
+    results = Fido.search(...)
+    files = Fido.fetch(results, path=path)
+
 
 @pytest.mark.remote_data
-@given(time_attr(time=datetimes(timezones=[], max_year=datetime.datetime.utcnow().year, min_year=2010)))
+@given(time_attr(time=datetimes(timezones=[], 
+    max_year=datetime.datetime.utcnow().year, min_year=2010)))
```
Comment (nabobalis) [inline sunpy/net/dataretriever/tests/test_eve.py]: `max_year` should be inline, vertically,  with `time_attr`

File: sunpy/net/fido_factory.py [hunk 6]
```diff
@@ -327,7 +328,7 @@ def search(self, *query):
 
     # Python 3: this line should be like this
     # def fetch(self, *query_results, wait=True, progress=True, **kwargs):
-    def fetch(self, *query_results, **kwargs):
+    def fetch(self, *query_results, **kwargs, path=""):
```
Comment (nabobalis) [inline sunpy/net/fido_factory.py]: Order is wrong here.

Comment (nabobalis) [reply sunpy/net/fido_factory.py]: Als should be `'` instead of `"`

File: sunpy/net/fido_factory.py [hunk 7]
```diff
@@ -350,16 +351,22 @@ def fetch(self, *query_results, **kwargs):
...
-            for block in query_result.responses:
-                reslist.append(block.client.fetch(block, **kwargs))
+        if path is None:
+            for query_result in query_results:
+                for block in query_result.responses:
+                    reslist.append(block.client.fetch(block, **kwargs))
+        else:
+            for query_result in query_results:
+                for block in query_result.responses:
+                    reslist.append(block.client.fetch(block, **kwargs, path=path))
```
Comment (nabobalis) [inline sunpy/net/fido_factory.py]: This `if,else` seems overkill since we are copying the same code twice just adding in a path keyword. 

It would be better to handle a path to None and handle that.

**Comment by nabobalis** [comment]
Also once you have the changes made, can you provide a short code example and output please.

File: sunpy/database/database.py (hunk 1 above)
Comment (Cadair) [inline sunpy/database/database.py]: why this change at all?

File: sunpy/database/database.py (hunk 2 above)
Comment (Cadair) [reply sunpy/database/database.py]: I don't understand the reasoning for this, can you explain @jnanjekye

**Comment by jnanjekye** [comment]
@nabobalis @Cadair Thanks for the review ..I will get through a recent commit that has addressed the above and has no [WIP] label. I had commited a patch with [WIP] label to start ground so I could get help on the error I was getting while running tests.

I am now able to run my tests and will soon commit a much better version of this patch.

**Comment by nabobalis** [comment]
I removed it in the title and added it as a label. The labels are more for me and review purpose then anything else.

I think a test would be good as well to make sure the patch works.

**Comment by nabobalis** [comment]
@jnanjekye Any chance of an update on this?

**Comment by jnanjekye** [comment]
Thanks for the reminder...earlier tommorrow.

**Comment by jnanjekye** [comment]
@vn-ki please do I will close this so that you open a new PR.
